src/main/java/com/iggroup/universityworkshopmw/domain/services/OpenPositionsService.java

```diff
@@ -0,0 +1,128 @@
+import com.iggroup.universityworkshopmw.domain.model.Client;
+import com.iggroup.universityworkshopmw.domain.model.OpenPosition;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import static com.google.common.collect.Lists.newArrayList;
+import static com.iggroup.universityworkshopmw.domain.helpers.Helper.createUniqueId;
+
+@Slf4j
+@Component
+public class OpenPositionsService {
+   private final ClientService clientService;
+   private ConcurrentMap<String, List<OpenPosition>> clientPositionStore;
+
+   public OpenPositionsService(ClientService clientService) {
+      this.clientService = clientService;
+      clientPositionStore = new ConcurrentHashMap<>();
+
+   }
+
+   public List<OpenPosition> getOpenPositionsForClient(String clientId) throws NoAvailableDataException {
+      List<OpenPosition> openPositions = getPositionDataFromMap(clientId);
+
+      if (openPositions != null) {
+         return openPositions;
+      }
+      throw new NoAvailableDataException("No open positions exist for client: " + clientId);
+   }
+
+   public void addOpenPositionForClient(String clientId, OpenPosition newOpenPosition) throws NoAvailableDataException, InsufficientFundsException {
+      double positionPrice = newOpenPosition.getBuySize() * newOpenPosition.getOpeningPrice();
+      List<OpenPosition> openPositionsForClient = clientPositionStore.get(clientId);
+
+      double clientFunds = checkClientFunds(clientId, positionPrice);
+      updateStoreWithNewPosition(clientId, newOpenPosition, openPositionsForClient);
+
+      clientService.updateFunds(clientId, clientFunds);
+   }
+
+   public Double closeOpenPosition(String clientId, String openPositionToClose, Double closingPrice) throws NoAvailableDataException {
+      List<OpenPosition> openPositions = getPositionDataFromMap(clientId);
+
+      OpenPosition position = openPositions.stream()
+         .filter(pos -> pos.getId().equals(openPositionToClose))
+         .findFirst()
+         .orElseThrow(() -> new NoAvailableDataException("No position exists with id: " + openPositionToClose));
+
+      Double closingProfitAndLoss = calculateNewProfitAndLoss(closingPrice, position.getOpeningPrice(), position.getBuySize());
+
+      openPositions.remove(position);
+      if (openPositions.isEmpty()) {
+         clientPositionStore.remove(clientId);
+      }
+
+      updateClientFunds(clientId, closingProfitAndLoss);
+      return closingProfitAndLoss;
+   }
+
+   public void updateMarketPrice(String marketId, Double newValue) {
+      clientPositionStore.keySet().stream()
+         .forEach(client -> {
+            List<OpenPosition> openPositions = clientPositionStore.get(client);
+            openPositions.stream()
+               .forEach(openPosition -> updateProfitAndLoss(marketId, newValue, openPosition, openPositions));
```

Comment:
we can add a filter here for positions for the marketId then no need to have the if statement in updateProfitAndLoss